src/parquet/file/reader.cc

```diff
@@ -74,6 +74,18 @@ ParquetFileReader::~ParquetFileReader() {
   Close();
 }
 
+std::unique_ptr<ParquetFileReader> ParquetFileReader::OpenStream(
+    ExternalInputStream* stream, MemoryAllocator* allocator) {
```

Comment:
From the `ExternalInputStream` abstract class:

```c++
  // Read interface to read from stream at an offset into a specified buffer
   // Returns the number of bytes read.
   virtual int64_t Read(int64_t num_to_read, int64_t offset, uint8_t* buffer) = 0;
```

I may be missing something, but this looks like random access to me. I do not see why `ExternalInputStream` or `StreamSource` are needed if the user can implement their own custom subclass of `RandomAccessSource` in downstream code.

So long as `ParquetFileReader` depends on being able to seek to the end of the file and read the metadata from the footer, all input sources need to support random access. There may be a case where the metadata is read separately from the rest of the file (which is streamed, so seeking is not supported).